### DISCUSSION TITLE
Subscribe to topic with callback

### DISCUSSION THREAD

**Discussion by sude22** [bertmelis/espMqttClient#78] [General]
Hi,
I'm not sure if I missed something or if the idea is fundamentally wrong...
Would it be a good feature if you could specify a callback when subscribing to a topic that would be called when the message is received?

**Comment by bertmelis** [comment]
A callback per subscribed topic? I understand the need but I probably won't add this to the library.
I might add an example on how to do this in the docs (in the [code samples section](https://github.com/bertmelis/espMqttClient/blob/main/docs/index.md#code-samples) but it isn't trivial for wildcard topics.

**Reply by sude22** [reply]
OK, so I didn't overlook this function.
A example how you would solve that would be cool (for me, wildcard topics are not used).

**Reply by bertmelis** [reply]
I would do something like this. Mind that I kept it very simple without usable callback function signatures. It runs on Linux.

```cpp
#include <iostream>
#include <map>
#include <cstring>

// callback-per-topic signature
typedef void (*topicCallback)();

// topic matching function
struct MatchTopic {
  bool operator()(const char* a, const char* b) const {
    return strcmp(a, b) < 0;
  }
};

// map with topic/callback pairs
std::map<const char*, topicCallback, MatchTopic> topicCallbacks;

// callbacks per topic
void onTopic1() {
  std::cout << "Received a message on topic 1" << std::endl;
}
void onTopic2() {
  std::cout << "Received a message on topic 1" << std::endl;
}

// general callback
void generalCallback(const char* topic) {
  if (!topic) return;

  // search for matching topic
  auto it = topicCallbacks.find(topic);
  if (it != topicCallbacks.end()) {
    // if found, run specific callback
    (*it->second)();
  } else {
    // or log an error
    std::cout << "Topic didn't match" << std::endl;
  }
}


int main() {

  // fill the map with topic/callback pairs
  topicCallbacks.emplace("base/topic1", &onTopic1);
  topicCallbacks.emplace("base/topic2", &onTopic2);

  generalCallback("base/topic1");
  generalCallback("base/topic2");
  generalCallback("base/topic3");

  return 0;
}
```

https://onlinegdb.com/kPFj4tq-HC

**Reply by bertmelis** [reply]
I created a std::map with pairs of `const char*` and function pointers with a custom comparator (struct `MatchTopic`).

You can fill this map with topic-function pairs.

Then, in the general callback function, the one you pass the espMqttClient, you use this map to search for a matching topic in the map and call the accompanying function pointer.

To make this usable you will have to update the function signatures.
If you want to use wilddcard topics, you will have to add some magic to the `MatchTopic` method.

**Only the first match will be executed.**

**Comment by sude22** [comment]
Wow, thanks!
After all this years with ESP8266/Homie I'm still not very good using C++, but with this hints I definitly give it a try.

**Reply by bertmelis** [reply]
only practice... 5 years ago I was still writing led-blink examples.

**Comment by bertmelis** [comment]
Added the more suitable code sample to the docs: https://www.emelis.net/espMqttClient/#code-samples
